src/vault/contract.rs

```diff
@@ -161,13 +167,32 @@ impl VaultCovenant {
         let nums_key = XOnlyPublicKey::from_slice(point.to_xonly_bytes().as_slice())?;
         let secp = Secp256k1::new();
         Ok(TaprootBuilder::new()
-            .add_leaf(1, vault_trigger_withdrawal())?
-            .add_leaf(2, vault_complete_withdrawal(self.timelock_in_blocks))?
-            .add_leaf(2, vault_cancel_withdrawal())?
+            .add_leaf(1, vault_trigger_withdrawal(self.x_only_public_key()))?
+            .add_leaf(2, vault_complete_withdrawal(self.x_only_public_key(), self.timelock_in_blocks))?
+            .add_leaf(2, vault_cancel_withdrawal(self.x_only_public_key()))?
             .finalize(&secp, nums_key)
             .expect("finalizing taproot spend info with a NUMS point should always work"))
     }
 
+    fn x_only_public_key(&self) -> XOnlyPublicKey {
+        return self.keypair.unwrap().x_only_public_key().0
```

Comment:
since the VaultCovenant structure has keypair as an `Option`, this should probably return an Option or we should require that VaultCovenant has a keypair

Reply:
I have updated it so that keypair is now required for VaultCovenant